charms/slurmd/src/utils/machine.py

```diff
@@ -0,0 +1,32 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Query information about the underlying Juju machine."""
+
+import logging
+import subprocess
+from typing import Dict
+
+_logger = logging.getLogger(__name__)
+
+
+def get_slurmd_info() -> Dict[str, str]:
+    """Get machine info as reported by `slurmd -C`."""
+    try:
+        r = subprocess.check_output(["slurmd", "-C"], text=True).strip()
+    except subprocess.CalledProcessError as e:
+        _logger.error(e)
+        raise
+
+    return dict([opt.split("=") for opt in r.split()[:-1]])
```

Comment:
Thought: Possibly something that we should move to `slurm_ops`?

Reply:
I wouldn't be against moving this logic into `slurm_ops`, but I think it's out of scope for this PR. My main reasoning is I'd like to further scope out what the "feature detection" story will be on the compute nodes so that we have a better idea how we'll identify the capabilities of each compute node.

I'd like to consider "resurrecting" [`sysprober`](https://github.com/charmed-hpc/sysprober) to see if it can provide similar functionality to `slurmd -C` - _they read the same files at the end of the day_ - because that way we can bundle together all our logic for compute node enlistment, accelerator detection, virtualization detection, etc.